### PR TITLE
EKA-142 | Acts on notification from consentmanager and hip when conse…

### PR DESCRIPTION
**Pull request by Veena-tw**
…nt status changes to expired

* listens to data-notification events and stores the data in a path of localstorage path, consent-request-id, transactionid, file
* upon consent expiry deletes the contents under the transaction id folder for that consent
* validates consent expiry prior to data flow initiation
* listens to consent-notification events, updates consent status, notifies hip, deletes local data
* appends the transaction-id to the attachments path

### CODE REVIEW

File: src/main/java/in/org/projecteka/hiu/HiuConfiguration.java [hunk 1]
```diff
@@ -15,21 +15,8 @@
...
-import in.org.projecteka.hiu.dataflow.DataFlowRequestListener;
-import in.org.projecteka.hiu.dataflow.DataFlowService;
-import in.org.projecteka.hiu.dataflow.DataFlowServiceProperties;
-import in.org.projecteka.hiu.dataflow.Decryptor;
-import in.org.projecteka.hiu.dataflow.HealthInfoManager;
-import in.org.projecteka.hiu.dataflow.HealthInformationRepository;
-import in.org.projecteka.hiu.dataflow.LocalDataStore;
+import in.org.projecteka.hiu.consent.*;
+import in.org.projecteka.hiu.dataflow.*;
 import in.org.projecteka.hiu.dataprocessor.DataAvailabilityListener;
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]: We chose to not use wild card imports for future name collisions.

Comment (Veena-tw) [reply src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]: changed it back

File: src/main/java/in/org/projecteka/hiu/ClientError.java [hunk 2]
```diff
@@ -41,6 +41,14 @@ public static ClientError consentArtefactNotFound() {
                 new ErrorRepresentation(new Error(CONSENT_ARTEFACT_NOT_FOUND, "Cannot find the consent artefact")));
     }
 
+    public static ClientError consentArtefactGone() {
+        return new ClientError(
+                HttpStatus.GONE,
+                new ErrorRepresentation(new Error(
+                        ErrorCode.CONSENT_ARTEFACT_NOT_FOUND,
+                        "Consent artefact expired")));
+    }
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/ClientError.java]: format it properly, follow the same pattern other in the same class.

File: src/main/java/in/org/projecteka/hiu/HiuConfiguration.java [hunk 3]
```diff
@@ -265,6 +277,24 @@ public DataFlowRequestListener dataFlowRequestListener(
...
     }
 
+    @Bean
+    public DataFlowDeleteListener dataFlowDeleteListener(
+            MessageListenerContainerFactory messageListenerContainerFactory,
+            DestinationsConfig destinationsConfig,
+            DataFlowRepository dataFlowRepository,
+            HealthInformationRepository healthInformationRepository,
+            DataFlowServiceProperties dataFlowServiceProperties,
+            LocalDataStore localDataStore
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]:
```suggestion
            LocalDataStore localDataStore)
```

File: src/main/java/in/org/projecteka/hiu/HiuConfiguration.java [hunk 4]
```diff
@@ -265,6 +277,24 @@ public DataFlowRequestListener dataFlowRequestListener(
...
+    ) {
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/HiuConfiguration.java]:
```suggestion
    {
```

File: src/main/java/in/org/projecteka/hiu/consent/ConsentService.java [hunk 5]
```diff
@@ -59,26 +60,29 @@
...
+                return validateRequest(consentNotificationRequest.getConsentRequestId())
+                        .filter(consentRequest -> consentRequest.getStatus() == REQUESTED)
+                        .switchIfEmpty(Mono.error(new ClientError(CONFLICT,
+                                new ErrorRepresentation(new Error(VALIDATION_FAILED,
+                                        "Consent request is already updated.")))))
+                        .flatMap(consentRequest ->
+                                consentRepository.updateConsent(consentRequest.getId(),
+                                        consentRequest.toBuilder().status(DENIED).build()));
+            default:
+                //TODO: Need to validate for all scenarios
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/consent/ConsentService.java]: you can remove this right as we covered all the status?

File: src/main/java/in/org/projecteka/hiu/consent/ConsentService.java [hunk 6]
```diff
@@ -103,20 +107,32 @@
...
+                                    consentNotificationRequest.getTimestamp());
+                        default:
+                            return processRejectedConsent(consentArtefactReference,
+                                    consentNotificationRequest.getStatus(),
+                                    consentNotificationRequest.getTimestamp());
+                    }
                 });
     }
 
+    private Mono<Void> processExpiredConsent(ConsentArtefactReference consentArtefactReference, String consentRequestId, ConsentStatus status, Date timestamp) {
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/consent/ConsentService.java]: line length violation.

File: src/main/java/in/org/projecteka/hiu/consent/DataFlowDeletePublisher.java [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+import reactor.core.publisher.Mono;
+
+import static in.org.projecteka.hiu.ClientError.queueNotFound;
+import static in.org.projecteka.hiu.HiuConfiguration.DATA_FLOW_DELETE_QUEUE;
+
+@AllArgsConstructor
+public class DataFlowDeletePublisher {
+    private static final Logger logger = LoggerFactory.getLogger(DataFlowDeletePublisher.class);
+    private AmqpTemplate amqpTemplate;
+    private DestinationsConfig destinationsConfig;
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/consent/DataFlowDeletePublisher.java]: private final

File: src/main/java/in/org/projecteka/hiu/consent/model/ConsentRequestRepresentation.java [hunk 8]
```diff
@@ -21,7 +21,7 @@ public static ConsentRequestRepresentation toConsentRequestRepresentation(
             Patient patient,
             in.org.projecteka.hiu.consent.model.ConsentRequest consentRequest) {
         var format = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss'Z'");
-        var withMillSeconds = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'Z'");
+        var withMillSeconds = new SimpleDateFormat("yyyy-MM-dd'T'HH:mm:ss.SSS'+0000'");
```
Comment (ganesanarun) [inline src/main/java/in/org/projecteka/hiu/consent/model/ConsentRequestRepresentation.java:24]: Why we are changing this here?

File: src/main/java/in/org/projecteka/hiu/ClientError.java (hunk 2 above)
Comment (Veena-tw) [reply src/main/java/in/org/projecteka/hiu/ClientError.java]: done

File: src/main/java/in/org/projecteka/hiu/consent/model/ConsentRequestRepresentation.java (hunk 8 above)
Comment (Veena-tw) [reply src/main/java/in/org/projecteka/hiu/consent/model/ConsentRequestRepresentation.java:24]: That's how the actual date is getting stored in string format

**Review by ganesanarun — APPROVED** [review]

**Pull request merged**